fpga_interchange/fasm_generators/xc7.py

```diff
@@ -85,6 +85,101 @@ def parse_lut_bel(lut_bel):
 
 
 class XC7FasmGenerator(FasmGenerator):
+    def handle_brams(self):
+        """
+        Handles slice RAMB18 FASM feature emission.
+        """
+
+        # TODO: handle ramb36 as well
+
+        init_re = re.compile("INITP?_[0-9][0-9]")
+
+        z_features = ["INIT_A", "INIT_B", "SRVAL_A", "SRVAL_B"]
+        str_features = [
+            "RDADDR_COLLISION_HWCONFIG", "RSTREG_PRIORITY_A",
+            "RSTREG_PRIORITY_B", "WRITE_MODE_A", "WRITE_MODE_B"
+        ]
+        rw_widths = [
+            "READ_WIDTH_A", "READ_WIDTH_B", "WRITE_WIDTH_A", "WRITE_WIDTH_B"
+        ]
+
+        allowed_cell_types = ["RAMB18E1"]
+        allowed_site_types = ["RAMB18E1"]
+
+        fasm_features = list()
+        for cell_instance, cell_data in self.physical_cells_instances.items():
+            cell_type = cell_data.cell_type
+            if cell_type not in allowed_cell_types:
+                continue
+
+            tile_name = cell_data.tile_name
+            tile_type = cell_data.tile_type
+            site_name = cell_data.site_name
+            bram_prefix = self.get_bram_prefix(site_name, tile_type)
+
+            is_y1 = "Y1" in bram_prefix
+
+            self.add_cell_feature((tile_name, bram_prefix, "IN_USE"))
+
+            attributes = cell_data.attributes
+
+            ram_mode = attributes["RAM_MODE"]
+            for attr, value in attributes.items():
+                init_param = self.device_resources.get_parameter_definition(
+                    cell_type, attr)
+
+                init_match = init_re.match(attr)
+                fasm_feature = None
+                if init_match:
+                    init_value = init_param.decode_integer(value)
+
+                    if init_value == 0:
+                        continue
+
+                    init_str_value = "{:b}".format(init_value)
+                    init_str = "{len}'b{value}".format(
+                        len=len(init_str_value), value=init_str_value)
+                    fasm_feature = "{}[{}:0]={}".format(
+                        attr, len(init_str_value), init_str)
+                    fasm_features.append(fasm_feature)
+
+                elif attr in z_features:
+                    init_value = init_param.decode_integer(value)
+                    width = init_param.width
+
+                    init_str_value = "{value:0{width}b}".format(
+                        value=init_value, width=width)
+
+                    feature_value = init_str_value.replace("1", "2").replace(
```

Comment:
I'm curious where the 2s are coming from? this seems to be a binary string

Reply:
Hmm.. I need to double check if replacing 1s with 2s is actually permitted in this case, but I think that there actually are neater ways to implement a bit string inversion.

Reply:
Oh, I missed that this was a trick for inversion.

If you want to keep using this trick, putting it in a function called something like `invert_bitstring` would make it a lot clearer.

Reply:
Yeah, I guess wrapping this in a function makes sense, as bitstring inversion might be something generally useful

Reply:
Done